editor/src/document/document_message_handler.rs

```diff
@@ -43,6 +44,52 @@ impl DocumentsMessageHandler {
 	pub fn active_document_mut(&mut self) -> &mut DocumentMessageHandler {
 		&mut self.documents[self.active_document_index]
 	}
+	fn get_new_doc_title(&self) -> String {
+		let digits = ('0'..='9').collect::<Vec<char>>();
+		let mut doc_title_numbers = self
+			.documents
+			.iter()
+			.map(|d| {
+				if d.name.ends_with(digits.as_slice()) {
+					let (_, number) = d.name.split_at(17);
+					number.trim().parse::<usize>().unwrap()
+				} else {
+					1
+				}
+			})
+			.collect::<Vec<usize>>();
+		doc_title_numbers.sort_unstable();
+		let mut new_doc_title_num = 1;
+		while new_doc_title_num <= self.documents.len() {
+			if new_doc_title_num != doc_title_numbers[new_doc_title_num - 1] {
+				break;
+			}
+			new_doc_title_num += 1;
+		}
```

Comment:
this could be probably made more readable by doing something like this:
```rust
let (new_num, _) = doc_title_numbers.enumerate().find(|(i, number)| i != number).unwrap_or(self.documents.len());
```

Reply:
I tried implementing this but i ran into some issues so I have left it as-is for now.

Reply:
Ah yeah, my bad, wanted to post this
```rust
let (_, new_num) = doc_title_numbers.enumerate().find(|(i, number)| i != number).unwrap_or((0,self.documents.len()));
```
I am not sure what you mean by It would also return 1.
Oh btw I think this could actually be made more efficient by using a binary search.
```rust
// Uses binary search to find the index of the element where number is bigger than i
let new_num = doc_title_numbers.enumerate().binary_search_by(|(i,number)|number.cmp(i + 1)).map_or_else(|e|e, |v|v) + 1;
```
I think something like this should work, I knew there was some kind of optimization hiding

Reply:
Oh and while it would be pretty cool to use a fancy bin search, this would be mostly to set a precedent and to write better code in the future (and inspire others reading the code to do the same). I don't think we will ever have enough elements for this to make any difference.
And I just realized this does not work because we operate on an iterator for which bin search is not implemented…
we would need:
```rust
(Make the vector consist of i32 somewhere above during the collection)
doc_title_numbers.iter_mut().enumerate.for_each(|(i, number)| number = number -i -1);
// Uses binary search to find the index of the element where number is bigger than i
let new_num = doc_title_numbers.binary_search(0).map_or_else(|e|e, |v|v) + 1;
```